torch/_inductor/decomposition.py

```diff
@@ -144,8 +167,138 @@ def addmm(input, mat1, mat2, *, beta=1, alpha=1):
+
+        if op is torch.ops.aten.mm or op is torch.ops.aten.addmm:
+            mat1_pad = pad_dim(mat1, get_padded_length(mat1.shape[1]), 1)
+            mat2_pad = mat2.new_empty([get_padded_length(i) + i for i in mat2.shape])
+            if op is torch.ops.aten.addmm:
+                if input is not None:
+                    if input.dim() == 2:
+                        input_pad = pad_dim(input, get_padded_length(input.shape[1]), 1)
+                    elif input.dim() == 1:
+                        input_pad = pad_dim(input, get_padded_length(input.shape[0]), 0)
+                pad_time = bench(lambda: op(input_pad, mat1_pad, mat2_pad))
+            else:
+                pad_time = bench(lambda: op(mat1_pad, mat2_pad))
+        elif op is torch.ops.aten.bmm:
+            mat1_pad = pad_dim(mat1, get_padded_length(mat1.shape[2]), 2)
+            mat2_pad = mat2.new_empty(
+                mat2.shape[0],
+                get_padded_length(mat2.shape[1]) + mat2.shape[1],
+                get_padded_length(mat2.shape[2]) + mat2.shape[2],
+            )
+            pad_time = bench(lambda: op(mat1_pad, mat2_pad))
+        else:
+            return False
+
+        # Shape padding introduces addtional memory ops. Based on microbenchmarks, 1.4x represents a reasonable
+        # tradeoff between performance improvement from shape padding and overhead from addtional memory ops
+        # TODO: Build a learned model which would be better than this heuristic
+        return ori_time > pad_time * 1.4
+
+
+@register_decomposition([aten.mm])
+def mm_decomp(mat1, mat2):
+    if (
+        config.shape_padding
+        and check_device_dtype(mat1, mat2)
+        and should_pad_bench(mat1, mat2, torch.ops.aten.mm)
+    ):
+
+        k_padded_length = get_padded_length(mat1.shape[1])
+        n_padded_length = get_padded_length(mat2.shape[1])
+
+        if k_padded_length != 0:
+            mat1 = pad_dim(mat1, k_padded_length, 1)
+            mat2 = pad_dim(mat2, k_padded_length, 0)
+        if n_padded_length != 0:
+            mat2 = pad_dim(mat2, n_padded_length, 1)
+            return torch.ops.aten.mm(mat1, mat2)[:, :-n_padded_length]
+        elif k_padded_length != 0:
+            return torch.ops.aten.mm(mat1, mat2)
+
+    return NotImplemented  # go directly to lowering
+
+
+@register_decomposition([aten.bmm])
+def bmm_decomp(mat1, mat2):
+    if (
+        config.shape_padding
+        and check_device_dtype(mat1, mat2)
+        and should_pad_bench(mat1, mat2, torch.ops.aten.bmm)
+    ):
+
+        k_padded_length = get_padded_length(mat1.shape[2])
+        n_padded_length = get_padded_length(mat2.shape[2])
+
+        if k_padded_length != 0:
+            mat1 = pad_dim(mat1, k_padded_length, 2)
+            mat2 = pad_dim(mat2, k_padded_length, 1)
+        if n_padded_length != 0:
+            mat2 = pad_dim(mat2, n_padded_length, 2)
+            return torch.ops.aten.bmm(mat1, mat2)[:, :, :-n_padded_length].contiguous()
```

Comment:
I added this contiguous call sinceI ran into needing it for correctness, but I think this might actually necessary in all cases?

cc: @ngimel 

It does sadly seem to impact perf a bit in my testing though.